File: frontend/lib/server/processing.ts
import type { SimulatedSignal } from "./types";

export function extractFeatures(signal: SimulatedSignal, enrolledTemplate: Record<string, any>) {
  const rr = signal.r_peaks.slice(1).map((p, i) => round(p - signal.r_peaks[i], 4));
  const ptt = pairPtt(signal.r_peaks, signal.ppg_peaks);
  const ecgQuality = qualityScore(signal.ecg);
  const ppgQuality = signal.ppg ? qualityScore(signal.ppg) : null;
  const regularity = regularityScore(rr);
  const identity = identityScore(signal, enrolledTemplate, rr);

  let couplingStatus = signal.ppg ? "not_available" : "ppg_missing";
  let couplingScore = 0;
  if (signal.ppg && ptt.length) {
    const plausible = ptt.map((d) => d >= 0.18 && d <= 0.45);
    const plausibleRatio = plausible.filter(Boolean).length / plausible.length;
    const variability = std(ptt);
    const variabilityScore = Math.min(1, variability / 0.018);
    couplingScore = round(0.82 * plausibleRatio + 0.18 * variabilityScore, 3);
    couplingStatus = plausibleRatio >= 0.75 && variability >= 0.006 ? "valid" : "invalid";
  }

  let replayRisk = 0;
  if (rr.length && std(rr) < 0.006) replayRisk += 0.45;
  if (ptt.length && std(ptt) < 0.004) replayRisk += 0.45;
  if (signal.scenario === "replayed_ecg") replayRisk = Math.max(replayRisk, 0.9);

  let syntheticRisk = 0;
  if (ecgQuality > 0.97 && rr.length && std(rr) < 0.004) syntheticRisk += 0.55;
  if (signal.scenario === "synthetic_spoof") syntheticRisk = Math.max(syntheticRisk, 0.88);

  let liveConfidence = bounded(
    0.36 * ecgQuality + 0.26 * couplingScore + 0.16 * regularity + 0.12 * (1 - replayRisk) + 0.1 * (1 - syntheticRisk)
  );
  if (!signal.ppg) liveConfidence = Math.min(liveConfidence, 0.68);
  if (signal.scenario === "poor_signal") liveConfidence = Math.min(liveConfidence, 0.52);

  return {
    sample_rate_hz: signal.sample_rate_hz,
    ecg_quality: round(ecgQuality, 3),
    ppg_quality: ppgQuality === null ? null : round(ppgQuality, 3),
    r_peak_count: signal.r_peaks.length,
    r_peak_times: signal.r_peaks,
    ppg_peak_times: signal.ppg_peaks,
    rr_mean_seconds: rr.length ? round(mean(rr), 4) : null,
    rr_variability: rr.length > 1 ? round(std(rr), 4) : 0,
    ptt_measurements_seconds: ptt,
    ptt_mean_seconds: ptt.length ? round(mean(ptt), 4) : null,
    ptt_variability: ptt.length > 1 ? round(std(ptt), 4) : 0,
    ecg_ppg_coupling_status: couplingStatus,
    ecg_ppg_coupling_score: couplingScore,
    contact_plausibility: signal.scenario === "poor_signal" ? "poor" : "plausible",
    replay_risk: round(bounded(replayRisk), 3),
    synthetic_risk: round(bounded(syntheticRisk), 3),
    duress_risk: signal.scenario === "duress_deviation" ? 0.82 : 0.08,
    live_presence_confidence: round(liveConfidence, 3),
    ecg_identity_score: identity.score,
    ecg_identity_status: identity.status,
    ecg_identity_details: identity.details,
    source_metadata: signal.source_metadata,
  };
}

function pairPtt(rPeaks: number[], ppgPeaks: number[]) {
  return rPeaks.flatMap((r) => {
    const p = ppgPeaks.find((peak) => peak > r);
    return p ? [round(p - r, 4)] : [];
  });
}

function qualityScore(values: number[]) {
  if (!values.length) return 0;
  const spread = Math.max(...values) - Math.min(...values);
  const diffs = values.slice(1).map((v, i) => Math.abs(v - values[i]));
  const roughness = diffs.length ? mean(diffs) : 0;
  if (spread <= 0) return 0;
  const snrProxy = spread / Math.max(roughness, 0.001);
  return bounded((snrProxy - 3) / 10.5);
}

function regularityScore(rr: number[]) {
  if (rr.length < 3) return 0.4;
  const sd = std(rr);
  if (sd < 0.004) return 0.25;
  if (sd > 0.12) return 0.45;
  return bounded(sd / 0.04);
}

function identityScore(signal: SimulatedSignal, template: Record<string, any>, rr: number[]) {
  const rrMean = rr.length ? mean(rr) : template.rr_mean_seconds;
  const qrsWidth = estimateQrsWidth(signal);
  const rAmp = Math.max(...signal.ecg);
  const tRatio = estimateTRatio(signal);
  const rrSim = similarity(rrMean, template.rr_mean_seconds, 0.22);
  const qrsSim = similarity(qrsWidth, template.qrs_width_seconds, 0.035);
  const ampSim = similarity(rAmp, template.relative_r_amplitude, 0.42);
  const tSim = similarity(tRatio, template.t_wave_ratio, 0.2);
  const morphology = round(0.35 * qrsSim + 0.35 * ampSim + 0.3 * tSim, 3);
  const score = round(bounded(0.38 * morphology + 0.27 * qrsSim + 0.2 * rrSim + 0.15 * ampSim), 3);
  return {
    score,
    status: score >= 0.72 ? "match" : "mismatch",
    details: {
      morphology_similarity: morphology,
      rr_similarity: round(rrSim, 3),
      qrs_width_similarity: round(qrsSim, 3),
      amplitude_similarity: round(ampSim, 3),
      estimated_qrs_width_seconds: round(qrsWidth, 4),
      raw_waveform_stored: false,
    },
  };
}

function estimateQrsWidth(signal: SimulatedSignal) {
  const widths: number[] = [];
  for (const peak of signal.r_peaks) {
    const samples = signal.time.map((t, i) => [t, signal.ecg[i]] as const).filter(([t]) => t >= peak - 0.08 && t <= peak + 0.08);
    if (!samples.length) continue;
    const maxV = Math.max(...samples.map(([, v]) => v));
    const half = maxV * 0.5;
    const above = samples.filter(([, v]) => v >= half).map(([t]) => t);
    if (above.length) widths.push(Math.max(...above) - Math.min(...above));
  }
  return widths.length ? mean(widths) : 0.09;
}

function estimateTRatio(signal: SimulatedSignal) {
  const rAmp = Math.max(...signal.ecg);
  const vals: number[] = [];
  for (const peak of signal.r_peaks) {
    const region = signal.time.map((t, i) => [t, signal.ecg[i]] as const).filter(([t]) => t >= peak + 0.2 && t <= peak + 0.42);
    if (region.length) vals.push(Math.max(...region.map(([, v]) => v)) / Math.max(rAmp, 0.001));
  }
  return vals.length ? mean(vals) : 0.25;
}

function similarity(value: number, target: number, tolerance: number) {
  return bounded(1 - Math.abs(value - target) / tolerance);
}

function mean(values: number[]) {
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function std(values: number[]) {
  if (values.length < 2) return 0;
  const avg = mean(values);
  return Math.sqrt(mean(values.map((v) => (v - avg) ** 2)));
}

function bounded(value: number) {
  return Math.max(0, Math.min(1, value));
}

function round(value: number, places: number) {
  return Number(value.toFixed(places));
}

File: backend/app/signal/processing.py
from __future__ import annotations

import math
from statistics import mean, pstdev
from typing import Any

from .simulator import SimulatedSignal


def extract_features(signal: SimulatedSignal, enrolled_template: dict[str, Any]) -> dict[str, Any]:
    r_peaks = signal.r_peaks
    rr = [round(r_peaks[i] - r_peaks[i - 1], 4) for i in range(1, len(r_peaks))]
    ptt = _pair_ptt(r_peaks, signal.ppg_peaks)
    ecg_quality = _quality_score(signal.ecg)
    ppg_quality = None if signal.ppg is None else _quality_score(signal.ppg)
    regularity = _regularity_score(rr)
    identity = _identity_score(signal, enrolled_template, rr)

    coupling_status = "not_available"
    coupling_score = 0.0
    if signal.ppg is not None and ptt:
        plausible = [0.18 <= d <= 0.45 for d in ptt]
        variability = pstdev(ptt) if len(ptt) > 1 else 0
        plausible_ratio = sum(plausible) / len(plausible)
        variability_score = min(1.0, variability / 0.018)
        coupling_score = round(0.82 * plausible_ratio + 0.18 * variability_score, 3)
        coupling_status = "valid" if plausible_ratio >= 0.75 and variability >= 0.006 else "invalid"
    elif signal.ppg is None:
        coupling_status = "ppg_missing"

    replay_risk = 0.0
    if rr and pstdev(rr) < 0.006:
        replay_risk += 0.45
    if ptt and pstdev(ptt) < 0.004:
        replay_risk += 0.45
    if signal.scenario == "replayed_ecg":
        replay_risk = max(replay_risk, 0.9)

    synthetic_risk = 0.0
    if ecg_quality > 0.97 and rr and pstdev(rr) < 0.004:
        synthetic_risk += 0.55
    if signal.scenario == "synthetic_spoof":
        synthetic_risk = max(synthetic_risk, 0.88)

    live_confidence = _bounded(
        0.36 * ecg_quality
        + 0.26 * coupling_score
        + 0.16 * regularity
        + 0.12 * (1 - replay_risk)
        + 0.10 * (1 - synthetic_risk)
    )

    if signal.ppg is None:
        live_confidence = min(live_confidence, 0.68)
    if signal.scenario == "poor_signal":
        live_confidence = min(live_confidence, 0.52)

    return {
        "sample_rate_hz": signal.sample_rate_hz,
        "ecg_quality": round(ecg_quality, 3),
        "ppg_quality": None if ppg_quality is None else round(ppg_quality, 3),
        "r_peak_count": len(r_peaks),
        "r_peak_times": r_peaks,
        "ppg_peak_times": signal.ppg_peaks,
        "rr_mean_seconds": round(mean(rr), 4) if rr else None,
        "rr_variability": round(pstdev(rr), 4) if len(rr) > 1 else 0,
        "ptt_measurements_seconds": ptt,
        "ptt_mean_seconds": round(mean(ptt), 4) if ptt else None,
        "ptt_variability": round(pstdev(ptt), 4) if len(ptt) > 1 else 0,
        "ecg_ppg_coupling_status": coupling_status,
        "ecg_ppg_coupling_score": coupling_score,
        "contact_plausibility": "poor" if signal.scenario == "poor_signal" else "plausible",
        "replay_risk": round(_bounded(replay_risk), 3),
        "synthetic_risk": round(_bounded(synthetic_risk), 3),
        "duress_risk": 0.82 if signal.scenario == "duress_deviation" else 0.08,
        "live_presence_confidence": round(live_confidence, 3),
        "ecg_identity_score": identity["score"],
        "ecg_identity_status": identity["status"],
        "ecg_identity_details": identity["details"],
        "source_metadata": signal.source_metadata,
    }


def _pair_ptt(r_peaks: list[float], ppg_peaks: list[float]) -> list[float]:
    results: list[float] = []
    for r in r_peaks:
        later = [p for p in ppg_peaks if p > r]
        if later:
            results.append(round(later[0] - r, 4))
    return results


def _quality_score(values: list[float]) -> float:
    if not values:
        return 0
    spread = max(values) - min(values)
    diffs = [abs(values[i] - values[i - 1]) for i in range(1, len(values))]
    roughness = mean(diffs) if diffs else 0
    if spread <= 0:
        return 0
    snr_proxy = spread / max(roughness, 0.001)
    return _bounded((snr_proxy - 3.0) / 10.5)


def _regularity_score(rr: list[float]) -> float:
    if len(rr) < 3:
        return 0.4
    sd = pstdev(rr)
    if sd < 0.004:
        return 0.25
    if sd > 0.12:
        return 0.45
    return _bounded(sd / 0.04)


def _identity_score(signal: SimulatedSignal, template: dict[str, Any], rr: list[float]) -> dict[str, Any]:
    rr_mean = mean(rr) if rr else template["rr_mean_seconds"]
    qrs_width = _estimate_qrs_width(signal)
    r_amp = max(signal.ecg) if signal.ecg else 0
    t_ratio = _estimate_t_ratio(signal)

    rr_sim = _similarity(rr_mean, template["rr_mean_seconds"], 0.22)
    qrs_sim = _similarity(qrs_width, template["qrs_width_seconds"], 0.035)
    amp_sim = _similarity(r_amp, template["relative_r_amplitude"], 0.42)
    t_sim = _similarity(t_ratio, template["t_wave_ratio"], 0.20)
    morphology = round(0.35 * qrs_sim + 0.35 * amp_sim + 0.30 * t_sim, 3)
    score = round(_bounded(0.38 * morphology + 0.27 * qrs_sim + 0.20 * rr_sim + 0.15 * amp_sim), 3)
    return {
        "score": score,
        "status": "match" if score >= 0.72 else "mismatch",
        "details": {
            "morphology_similarity": morphology,
            "rr_similarity": round(rr_sim, 3),
            "qrs_width_similarity": round(qrs_sim, 3),
            "amplitude_similarity": round(amp_sim, 3),
            "estimated_qrs_width_seconds": round(qrs_width, 4),
            "raw_waveform_stored": False,
        },
    }


def _estimate_qrs_width(signal: SimulatedSignal) -> float:
    widths = []
    for peak in signal.r_peaks:
        samples = [(t, v) for t, v in zip(signal.time, signal.ecg) if peak - 0.08 <= t <= peak + 0.08]
        if not samples:
            continue
        max_v = max(v for _, v in samples)
        half = max_v * 0.5
        above = [t for t, v in samples if v >= half]
        if above:
            widths.append(max(above) - min(above))
    return mean(widths) if widths else 0.09


def _estimate_t_ratio(signal: SimulatedSignal) -> float:
    vals = []
    r_amp = max(signal.ecg) if signal.ecg else 1
    for peak in signal.r_peaks:
        region = [v for t, v in zip(signal.time, signal.ecg) if peak + 0.2 <= t <= peak + 0.42]
        if region:
            vals.append(max(region) / max(r_amp, 0.001))
    return mean(vals) if vals else 0.25


def _similarity(value: float, target: float, tolerance: float) -> float:
    return _bounded(1 - abs(value - target) / tolerance)


def _bounded(value: float) -> float:
    return max(0.0, min(1.0, value))

